core/dataset.cpp

```diff
@@ -884,6 +893,15 @@ Dataset &Dataset::operator/=(const Dataset &other) {
   return apply(divide_equals, *this, other);
 }
 
+Dimensions Dataset::dimensions() const {
+  Dimensions all;
+  for (const auto &dim : this->m_dims) {
+    auto extent = extents::decodeExtent(dim.second);
+    all.add(dim.first, extent); // TODO this is add outer rather than inner.
+  }
+  return all;
+}
+
```

Comment:
See other comment. We cannot use `Dimensions` as a return type when implementing such a method for `Dataset`.